src/indexer.rs

```diff
@@ -128,17 +129,38 @@ impl Indexer {
         let last_block = self.data.lock().unwrap().last_block;
         println!("Starting height: {}", last_block);
         let mut handle = None;
-        let (tx, mut rx) = tokio::sync::mpsc::channel(1);
+        let (tx1, mut rx1) = tokio::sync::mpsc::channel(2);
+        let (tx2, mut rx2) = tokio::sync::mpsc::channel(2);
+        let (tx3, mut rx3) = tokio::sync::mpsc::channel(2);
+        let (tx4, mut rx4) = tokio::sync::mpsc::channel(2);
 
         tokio::spawn(async move {
-            let _ = tokio::signal::ctrl_c().await;
-            let _ = tx.send(()).await;
+            let _t = tokio::signal::ctrl_c().await;
+            let _ = tx1.send(()).await;
+        });
+        tokio::spawn(async move {
+            let mut stream = tokio::signal::unix::signal(SignalKind::terminate()).unwrap();
+            let _t = stream.recv().await;
+            let _ = tx2.send(()).await;
+        });
+        tokio::spawn(async move {
+            let mut stream = tokio::signal::unix::signal(SignalKind::interrupt()).unwrap();
+            let _t = stream.recv().await;
+            let _ = tx3.send(()).await;
+        });
+        tokio::spawn(async move {
+            let mut stream = tokio::signal::unix::signal(SignalKind::quit()).unwrap();
+            let _t = stream.recv().await;
+            let _ = tx4.send(()).await;
```

Comment:
To be honest it looks not good. I'd create a new function `shutdown_listener`: 
```rust
fn shutdown_listener() -> tokio::sync::mpsc::Receiver<()> {
    use tokio::signal;

    let (tx, rx) = tokio::sync::mpsc::channel(1);
    let mut terminate = signal::unix::signal(SignalKind::terminate()).unwrap();
    let mut interrupt = signal::unix::signal(SignalKind::interrupt()).unwrap();
    let mut quit = signal::unix::signal(SignalKind::quit()).unwrap();

    tokio::spawn(async move {
        tokio::select! {
            _ = signal::ctrl_c() => tx.send(()).await,
            _ = terminate.recv() => tx.send(()).await,
            _ = interrupt.recv() => tx.send(()).await,
            _ = quit.recv() => tx.send(()).await,
        }
    });

    rx
}
```

and in method `run` 

```rust
       let mut shutdown_receiver = shutdown_listener();

        loop {
            tokio::select! {
                h = self.handle_block(&mut client) => handle = h,
                _ = shutdown_receiver.recv() => break,
                else => break,
            }
        }
```